sdk/src/resource/resource.cc

```diff
@@ -0,0 +1,74 @@
+#include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/sdk/resource/resource_detector.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+const std::string TELEMETRY_SDK_LANGUAGE = "telemetry.sdk.language";
+const std::string TELEMETRY_SDK_NAME     = "telemetry.sdk.name";
+const std::string TELEMETRY_SDK_VERSION  = "telemetry.sdk.version";
+
+Resource::Resource(const opentelemetry::common::KeyValueIterable &attributes) noexcept
+    : attribute_map_(attributes)
+{}
+
+Resource::Resource(const opentelemetry::sdk::trace::AttributeMap &attributes) noexcept
+    : attribute_map_(attributes)
+{}
+
+const opentelemetry::sdk::trace::AttributeMap &Resource::GetAttributes() const noexcept
+{
+  return attribute_map_;
+}
+
+std::unique_ptr<Resource> Resource::Merge(const Resource &other)
+{
+  opentelemetry::sdk::trace::AttributeMap merged_resource_attributes(other.GetAttributes());
+  merged_resource_attributes.AddAttributes(attribute_map_);
+  return std::unique_ptr<Resource>(new Resource(merged_resource_attributes));
+}
+
+std::unique_ptr<Resource> Resource::Create(
+    const opentelemetry::common::KeyValueIterable &attributes)
+{
+  Resource default_resource({{{TELEMETRY_SDK_LANGUAGE, "cpp"},
+                              {TELEMETRY_SDK_NAME, "opentelemetry"},
+                              {TELEMETRY_SDK_VERSION, OPENTELEMETRY_SDK_VERSION}}});
+
+  if (attributes.size() > 0)
+  {
+    Resource input_resource(attributes);
+    auto merged_resource = default_resource.Merge(input_resource);
+    return merged_resource->Merge(*(OTELResourceDetector().Detect()));
+  }
+  return default_resource.Merge(*(OTELResourceDetector().Detect()));
+}
+
+std::unique_ptr<Resource> Resource::CreateEmpty()
```

Comment:
I think when we start to use resources, we have to make them shared pointers. Resources are supposed to be attached to _every_ span. We don't want to have unique pointers there, but share a common resource with a shared pointer.

Reply:
Good point. Have changed it to use `shared_ptr`,

Reply:
Right. Have changed `std::unique_ptr` to `std::shared_ptr_ptr` now.

Reply:
Ok as per review comments from @jsuereth , Merge() and Create() now returns the stack allocated resource by value, and they resource is owned by `sdk::tracer_provider`, and passed as reference across `tracer` and `span`

Reply:
Ok. based on comments from @jsuereth , `Resources::Merge()` and `Resources::Create()` are modified to return stack allocated `Resource` by value. And this resource is then owned by `sdk::tracer_provider` and passed as reference to span. Please refer to `examples/simple/main.cc` in this PR on how the usage would be, and let me know if any concerns.